### DISCUSSION TITLE
Why spin instead of conventional web framework?

### DISCUSSION THREAD

**Discussion by mishushakov** [spinframework/spin#861] [Q&A] [answered]
Hey, I'd like to learn more about spin and one question I can't seem to find an answer for is: Why would I use spin instead of conventional web framework?

**Comment by dicej** [comment, marked as answer]
Hi @mishushakov.  There are a few reasons that come to mind:

1. Most conventional web frameworks (e.g. ExpressJS, Axum, etc.) are meant for stateful, long-lived services, where the handler for each request can read and modify shared, in-memory state within the service.  There are definitely advantages to that model, but Spin uses a "serverless" model instead, where each request handler runs in its own sandbox (similar to V8 isolates).  This kind of isolation provides much stronger security guarantees than a conventional framework, and is also much easier to scale horizontally, especially in combination with a distributed database like CockroachDB or PlanetScale.  Also, this isolation is achieved in a more portable and lightweight manner than containers and VMs.
2. Spin is meant to be used in tandem with the [Fermyon Platform](https://www.fermyon.dev/) and/or [Fermyon Cloud](https://cloud.fermyon.com/) as part of a frictionless, end-to-end development experience.  You can use Spin to test and debug your app locally and then deploy it to Platform or Cloud, either of which will take care of horizontal scaling, failover, etc.  Conventional web frameworks can be used in combination with e.g. Docker and Kubernetes to provide similar features, but the end-to-end development experience is significantly more complicated.
3. Since it's based on the WebAssembly Component Model, Spin can be used with a variety of programming languages and will soon allow you to efficiently and easily mix multiple programming languages in the same component.  The upshot is that you'll be able to integrate best-in-class libraries from a variety of language ecosystems rather than being restricted to just one of them.
